staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go

```diff
@@ -297,7 +301,10 @@ func NewCachedDiscoveryClientForConfig(config *restclient.Config, discoveryCache
 		return nil, err
 	}
 
-	return newCachedDiscoveryClient(discoveryClient, discoveryCacheDir, ttl), nil
+	// The delegate caches the discovery groups and resources (memcache). "ServerGroups",
+	// which usually only returns (and caches) the groups, can now store the resources as
+	// well if the server supports the newer aggregated discovery format.
+	return newCachedDiscoveryClient(memory.NewMemCacheClient(discoveryClient), discoveryCacheDir, ttl), nil
```

Comment:
This is confusing. As a requirement for beta, let's get down to a single level of caching against the new endpoint.  Future us will thank us for eliminating complexity here.

Reply:
Yes, why would we cache this at all on the client if they use the new format?

Reply:
First, let me name things for simplicity:
- "httpcache" is the smart-http cache, which in the case of discovery always needs to re-validate the cache against the server
- "diskcache" is a dummy disk cache, that saves on disk and doesn't know how to validate/invalidate
- "memcache" is a dummy memory cache, saves in memory for the lifespan of kubectl

Today, for discovery, kubectl uses diskcache, while controllers typically use the memcache. Everytime kubectl calls a function of the discovery, it reads the files from disk. We haven't implemented in this PR the diskcache for when aggregated exists, because it will use the httpcache instead. But then, it goes to the server to revalidate the cache every time (when doing `kubectl get pods` on a new cluster, it goes to the server 3 different times for no reason), so we're just using the memory cache too so that we don't reload from disk (or go to the server).

Reply:
Current Call Stack for `CachedDiscovery` calling `ServerGroupsAndResources` (or `ServerPreferredResources` or `ServerPreferredNamespacedResources`)
```
(CachedDiscovery) cd.ServerGroupsAndResources()
(DiscoveryClient) ServerGroupsAndResources(cd)
    cd.ServerGroups()
    (CachedDiscovery) ServerGroups()
        if not cached on disk (servergroups.json):
            delegate.ServerGroups()   <<<=== For aggregated discovery this returns Groups AND resources
    ...
    for each group/version retrieve resources:
        cd.ServerResourcesForGroupVersion()
        (CachedDiscovery) ServerResourcesForGroupVersion()
            if not cached on disk (serverresources.json):
                delegate.ServerResourcesForGroupVersion()   <<<=== A memcache delegate returns cached resources
```

Without a memcached delegate, we can not store the resources which are returned from aggregated discovery calling `ServerGroups`.

The alternative to a simple memcached delegate was significant complexity in the `CachedDiscoveryClient`.